### ISSUE TITLE
Add an `Exit condition` to the `Prevote` and `Precommit` steps when it is mathematically impossible to reach +2/3 weight on a single value

### ISSUE THREAD

**Issue by BrendanChou** [cometbft/cometbft#1430]
### Summary

Add an additional exit condition (importantly, without requiring timeout) to the `Prevote` or `Precommit` steps when it is clear that `+2/3 Pre{votes,commits}`, for `X` (whether `X` is a block or `nil`) is mathematically impossible.

Notably, this has two benefits:
- much faster resolution from several chain-halt scenarios
- faster time to finality (by moving through round steps faster) under non-deterministic `ProcessProposal` logic

### Problem Definition

Currently, we exit the `Prevote` or `Precommit` steps early (i.e. without requiring a timeout to be reached) once `+2/3 Pre{votes,commits}` are seen. That is, consensus is reached.

However, I would also argue that consensus is reached once it is clear that `+2/3 Pre{votes,commits}` is mathematically impossible to achieve. In this case, it is not necessary to wait for a timeout period and the validator can safely proceed to the next step as if `nil` is consensus.

Once it is true that the number of uncounted `Pre{votes,commits}` plus the mode (i.e. highest count) of the `Pre{votes,commits}` values sum to 2/3 or fewer of the total, then it is safe to say that the consensus value is `nil`.

This has two practical benefits which I mentioned above but will explain more deeply here.

Firstly, there is a faster time-to-resolution for a common chain-halt scenario where +2/3 of validators cannot come to agreement on the `AppHash`. Let us imagine the scenario where a non-determinism bug in `FinalizeBlock` results in 50% of validator weight having `AppHash=0x1` and 50% of validator weight having `AppHash=0x2`. This results in a chain-halt where rounds will increase as `Prevotes` never get +2/3 on any value and therefore all `Precommits` will be `nil`. At least 50% of the validators would have to patch their software, roll-back a single block, and then restart consensus from `Round 0`. However, if the current round got to a very high number, then it will take a very long time for such validators to get back to the highest round. This is because they will need to wait for the `Prevote` timeout period as no single value ever reached +2/3 weight.

Secondly, there is a faster time-to-finality for a blockchain that uses non-deterministic validation of blocks (e.g. `ProcessProposal`). Once it is certain that the currently proposed block will not reach +2/3 consensus, the network should instantly move-on to allow for a newly-proposed block. This would typically manifest itself in the `Prevote` step when 1/3+ of `Prevotes` are for the block `B` and 1/3+ of `Prevotes` are for `nil`. In the current algorithm, validators will still wait for the timeout period when this occurs.

### Proposal

Add an additional exit condition (not requiring a timeout) to the `Prevote` and `Precommit` steps [here](https://github.com/cometbft/cometbft/blob/98838143fe67f03f3d77f1ea1c28686d7cf0b155/spec/consensus/consensus.md?plain=1#L147-L196) when +2/3 weight on a single value becomes mathematically impossible.

**Comment by cason** [comment]
A rapid comment here: when you say that is "mathematically impossible" to reach a given number of votes for a given value, are you considering that Byzantine validators might vote for multiple values in a round step?

**Comment by BrendanChou** [comment]
While important to consider, I am thinking that is out-of-scope of this issue.

Let me know if you think that is fair or if that completely invalidates my suggestion.

**Comment by cason** [comment]
I don't think it invalidates the proposal. This is something that has been reported (e.g. https://github.com/tendermint/tendermint/issues/1496) and for which we should try to find a solution.

But to tell when it is impossible to get a quorum for a given value, you need to strictly define the level of fault-tolerance, namely the maximum voting power on the hands of Byzantine nodes. While Tendermint algorithm is designed for having <1/3 of Byzantine voting power, the implementation covers possible temporary situations on which this limit is not observed.

**Comment by cason** [comment]
A similar solution, which would have similar (or the same) effect, is to consider situations in which it doesn't make sense to wait for timeouts to expire, this is particular relevant for a recovering node. A workaround, which is actually possible, is to zero some of the timeouts when restarting a node, in the case of the first scenario. Then, once the critical height is committed, restart the node again to use the standard values for the timeout durations.

**Comment by cason** [comment]
The second scenario is a little different, if I understood it correctly. If I did, it aligns with some proposals we had in the past to early give up from a round that can't succeed at all. In this case, it is not just about removing an useless sleep time, but also skipping a round step (namely, from `prevote(R)` to `propose(R+1)`.

**Comment by BrendanChou** [comment]
> zero some of the timeouts when restarting a node

My engineering senses tells me that it's going to be a much better long-term solution to setting on a general solution rather than to break this up into two special cases.

As a side note, empirically we have seen that restarting the node to alter the timeouts results in starting from `Round=0`. We haven't had a chance to debug this. Originally it was our understanding that the WAL would hold this consensus information and not force a restart from `Round=0`.
